x-pack/plugins/encrypted_saved_objects/server/crypto/encryption_key_rotation_service.ts

```diff
@@ -0,0 +1,252 @@
+
+  /**
+   * The number of the Saved Objects that were still encrypted with one of the secondary encryption
+   * keys that we failed to re-encrypt with the primary key.
+   */
+  failed: number;
+}
+
+/**
+ * Service that deals with encryption key rotation matters.
+ */
+export class EncryptionKeyRotationService {
+  constructor(private readonly options: EncryptionKeyRotationServiceOptions) {}
+
+  public async rotate(
+    request: KibanaRequest,
+    { batchSize, type }: EncryptionKeyRotationParams
+  ): Promise<EncryptionKeyRotationResult> {
+    const [{ savedObjects }] = await this.options.getStartServices();
+    const typeRegistry = savedObjects.getTypeRegistry();
+
+    // We need to retrieve all SavedObject types which have encrypted attributes, specifically
+    // collecting those that are hidden as they are ignored by the Saved Objects client by default.
+    this.options.logger.debug('Retrieving Saved Object types that require encryption.');
+    const registeredSavedObjectTypes = [];
+    const registeredHiddenSavedObjectTypes = [];
+    for (const knownType of typeRegistry.getAllTypes()) {
+      if (this.options.service.isRegistered(knownType.name) && (!type || knownType.name === type)) {
+        registeredSavedObjectTypes.push(knownType.name);
+
+        if (knownType.hidden) {
+          registeredHiddenSavedObjectTypes.push(knownType.name);
+        }
+      }
+    }
+
+    const result = { total: 0, successful: 0, failed: 0 };
+    if (registeredSavedObjectTypes.length === 0) {
+      this.options.logger.debug(
+        type
+          ? `Saved Object type "${type}" is not registered, encryption key rotation is not needed.`
+          : 'There are no registered Saved Object types that can have encrypted attributes, encryption key rotation is not needed.'
+      );
+      return result;
+    }
+
+    this.options.logger.debug(
+      `Saved Objects with the following types [${registeredSavedObjectTypes}] will be processed.`
+    );
+
+    // We need two separate Saved Objects clients for the retrieval and update. For retrieval we
+    // don't want to have Encrypted Saved Objects wrapper so that it doesn't strip encrypted
+    // attributes. But for the update we want to have it so that it automatically re-encrypts
+    // attributes with the new primary encryption key.
+    const user = this.options.security?.authc.getCurrentUser(request) ?? undefined;
+    const retrieveClient = savedObjects.getScopedClient(request, {
+      includedHiddenTypes: registeredHiddenSavedObjectTypes,
+      excludedWrappers: ['encryptedSavedObjects'],
+    });
+    const updateClient = savedObjects.getScopedClient(request, {
+      includedHiddenTypes: registeredHiddenSavedObjectTypes,
+    });
+
+    // Keeps track of object IDs that have been processed already.
+    const processedObjectIDs = new Set<string>();
+
+    // Until we get scroll/search_after support in Saved Objects client we have to retrieve as much objects as allowed
```

Comment:
__note:__ in an absence of scroll/search_after support that's the only reasonable workaround I could come up with so far. Happy to hear other ideas, critique, objects, concerns! I tried to explain the approach in details in this comment, but let me know if something isn't clear.